R/functions.R

```diff
@@ -3,6 +3,34 @@
 #' Return id of the planning googlesheet
 gs_id <- function() "1HLtyGK_csi5W_v7XChxgTuVjS-RKXqc0Jxos1RBqpwk"
 
+
+#' Sanity checks for the google sheet.
+#' This function is used in data pipe-lines to check that everything
+#' in the google sheet is as expected.
+#' It will throw errors or warnings if something goes wrong.
+#'
+#' @details the function can be used in a dplyr pipe, which is why
+#' it requires and returns the `.data` argument.
+#' @param .data a data frame (or derivative) loaded from a google sheet
+sanity_checks <- function (.data) {
+  # https://github.com/hadley/assertthat/issues/41
+
+  # check if there are NA columns not supposed to be empty
+  required_columns <- c('Thema', 'Subthema', 'Start')
+  if (any(is.na(.data[, required_columns]))) {
+    troublemakers = rownames(.data)[rowSums(is.na(.data[, required_columns])) > 0]
+    message(paste0("Start and Deadline of a task may not be empty! (line/s ", troublemakers, ")")
+            )
```

Comment:
Or return the task codes rather than line numbers? (If they aren't missing :wink:). The row numbers in the data set are not the row numbers in the google sheet (difference of 1).

I'm not sure about the behaviour of `is.na()` on a data frame, i.e. `is.na(.data[, required_columns])`. From the documentation, it's not intended to do that.

I'd probably do something like:

```r
.data |> 
  filter(if_any({{required_columns}}, is.na)) |> 
  pull(taakcode)
```

Reply:
I got an
```
Error in `filter()`:
! Can't transform a data frame with duplicate names.
```
... and decided to resolve with base-R. It should work.
The data frame is not processed at the point of sanity check (which I think is the better way, for most checks, to mention hiccups on user entries asap).

Alternatively, we might move sanity checks down in the pipeline; but then we might miss issues cleared on the way.

Reply:
> `Can't transform a data frame with duplicate names.`

Right, so then better filter after the `clean_names()` step?

Reply:
> It should work.

Indeed you're right, `is.na()` seems to work for a data frame as well!

To get the row numbers, you could also simplify as: `which(rowSums(is.na(.data[, required_columns])) > 0)`

Reply:
- I moved the checks to after the `clean_names`, so that we have the option to use `dplyr` for checks.
- I also attempted the `filter` pipe you suggested above, but getting the `taakcode` did not work in my test case because that cell can and will likely be empty on erroneous lines (leaving an NA for warning).
- Warnings are indeed great, especially because of the `options(warn=...)` idea to optionally halt. Thank you for the book!
- Also thank you for reminding me of `which` and `pull`; many R functions still need some time and repetition to get stuck in my coding memory.

commit incoming.